File: Javascript/Stack & Queue/面试题0302.栈的最小值.js
/**
 * initialize your data structure here.
 */
var MinStack = function() {
    this.arr = []
    this.minArr = []
};

/** 
 * @param {number} x
 * @return {void}
 */
MinStack.prototype.push = function(x) {
    this.arr.push(x)
    if(x < this.minArr[this.minArr.length-1] || this.minArr.length == 0) {
        this.minArr.push(x)
    }else{
        this.minArr.push(this.minArr[this.minArr.length-1])
    }
};

/**
 * @return {void}
 */
MinStack.prototype.pop = function() {
    let temp = this.arr.splice(this.arr.length-1, 1)
    this.minArr.splice(this.minArr.length-1, 1)
    return temp
};

/**
 * @return {number}
 */
MinStack.prototype.top = function() {
    return this.arr[this.arr.length-1]
};

/**
 * @return {number}
 */
MinStack.prototype.getMin = function() {
    return this.minArr[this.minArr.length-1]
};

/**
 * Your MinStack object will be instantiated and called as such:
 * var obj = new MinStack()
 * obj.push(x)
 * obj.pop()
 * var param_3 = obj.top()
 * var param_4 = obj.getMin()
 */

File: Go/Stack/面试题 03.02. 栈的最小值.go
package Stack

// 双栈
type MinStack struct {
	primary 	[]int		// 主栈，存数据
	minVal		[]int		// 辅栈，存每次push后的最小值
}

func Constructor() MinStack {
	stack := new(MinStack)
	stack.primary = []int{}
	stack.minVal = []int{}
	return *stack
}

func (this *MinStack) Push(x int)  {
	this.primary = append(this.primary, x)
	min := x
	if len(this.minVal) > 0 {
		min = this.minVal[len(this.minVal)-1]
	}
	if x < min || 0 == len(this.minVal) {
		this.minVal = append(this.minVal, x)
	} else {
		this.minVal = append(this.minVal, min)
	}
}

func (this *MinStack) Pop()  {
	this.primary = this.primary[:len(this.primary)-1]
	this.minVal = this.minVal[:len(this.minVal)-1]
}

func (this *MinStack) Top() int {
	return this.primary[len(this.primary)-1]
}

func (this *MinStack) GetMin() int {
	return this.minVal[len(this.minVal)-1]
}